File: turf-reservation-frontend/src/api/booking.ts
import client from './client';

export interface BookingData {
    time_slot_id: number;
    total_price: number;
    coach_id?: number | null;
    event_id?: number | null;
}

export interface BookingResponse {
    booking_id: number;
    user_id: number;
    time_slot_id: number;
    coach_id?: number | null;
    event_id?: number | null;
    booking_date: string;
    status: 'confirmed' | 'cancelled' | 'pending';
    coach_approval_status: 'none' | 'pending' | 'approved' | 'rejected';
    admin_approval_status: 'none' | 'pending' | 'approved' | 'rejected';
    total_price: number;
    payment_status: 'pending' | 'paid';
    slot_date?: string;
    start_time?: string;
    end_time?: string;
    player_name?: string;
    player_email?: string;
    coach_name?: string;
    turf_name?: string;
}

export const bookingAPI = {
    createBooking: async (data: BookingData): Promise<BookingResponse> => {
        const response = await client.post<BookingResponse>('/api/bookings', data);
        return response.data;
    },

    getMyBookings: async (): Promise<BookingResponse[]> => {
        const response = await client.get<BookingResponse[]>('/api/bookings/my');
        return response.data;
    },

    rescheduleBooking: async (id: number, newTimeSlotID: number): Promise<{ message: string }> => {
        const response = await client.put<{ message: string }>(`/api/bookings/${id}/reschedule`, {
            new_time_slot_id: newTimeSlotID
        });
        return response.data;
    },

    cancelBooking: async (id: number): Promise<{ message: string }> => {
        const response = await client.post<{ message: string }>(`/api/bookings/${id}/cancel`, {});
        return response.data;
    },

    // Admin methods
    getPendingBookings: async (): Promise<BookingResponse[]> => {
        const response = await client.get<BookingResponse[]>('/api/admin/bookings/pending');
        return response.data;
    },

    approveBooking: async (id: number): Promise<{ message: string }> => {
        const response = await client.post<{ message: string }>(`/api/admin/bookings/${id}/approve`, {});
        return response.data;
    },

    rejectBooking: async (id: number): Promise<{ message: string }> => {
        const response = await client.post<{ message: string }>(`/api/admin/bookings/${id}/reject`, {});
        return response.data;
    },

    deleteBooking: async (id: number): Promise<{ message: string }> => {
        const response = await client.delete<{ message: string }>(`/api/bookings/${id}`);
        return response.data;
    },

    // Coach methods
    getCoachRequests: async (): Promise<BookingResponse[]> => {
        const response = await client.get<BookingResponse[]>('/api/coach/bookings');
        return response.data;
    },

    coachApproveBooking: async (id: number): Promise<{ message: string }> => {
        const response = await client.post<{ message: string }>(`/api/coach/bookings/${id}/approve`, {});
        return response.data;
    },

    coachRejectBooking: async (id: number): Promise<{ message: string }> => {
        const response = await client.post<{ message: string }>(`/api/coach/bookings/${id}/reject`, {});
        return response.data;
    }
};

export default bookingAPI;

File: turf-reservation-backend/internal/models/booking.go
package models

import (
	"time"
)

// Booking represents a reservation in the system
type Booking struct {
	BookingID     int       `json:"booking_id"`
	UserID        *int      `json:"user_id,omitempty"`
	TimeSlotID    *int      `json:"time_slot_id,omitempty"`
	CoachID       *int      `json:"coach_id,omitempty"`
	EventID       *int      `json:"event_id,omitempty"`
	BookingDate   time.Time `json:"booking_date"`
	Status              string    `json:"status"`
	CoachApprovalStatus string    `json:"coach_approval_status"`
	AdminApprovalStatus string    `json:"admin_approval_status"`
	TotalPrice          float64   `json:"total_price"`
	PaymentStatus       string    `json:"payment_status"`
	// Enriched fields from JOIN with time_slots
	SlotDate  string `json:"slot_date,omitempty"`
	StartTime string `json:"start_time,omitempty"`
	EndTime   string `json:"end_time,omitempty"`
	// Enriched fields from JOIN with users
	PlayerName  string `json:"player_name,omitempty"`
	PlayerEmail string `json:"player_email,omitempty"`
	CoachName   string `json:"coach_name,omitempty"`
	TurfName    string `json:"turf_name,omitempty"`
}
